[TAN-4519] Common Ground — BE foundations

adessy opened the pull request:
# Changelog
## Technical
- [TAN-4519] Implement the back-end ground work for the Common Ground participation method.

adessy committed:
[TAN-4519] Add the "common_ground_phase" factory

adessy committed:
[TAN-4519] Add documentation

adessy committed:
[TAN-4519] Implement minimal "Common Ground" participation method

adessy committed:
[TAN-4519] Define the default fields for the "Common Ground" form

For now, the form only includes a title field.

adessy committed:
[TAN-4519] Add type annotations

adessy committed:
[TAN-4519] Set default values for new inputs in Common Ground phases

adessy committed:
[TAN-4519] Test input creation in Common Ground phases

adessy committed:
[TAN-4519] Rework idea parameter defs for Common Ground phases

Common Ground inputs only need a title.

adessy committed:
[TAN-4519] Test Common Ground phase creation

adessy committed:
[TAN-4519] Test common_ground_phase factory

adessy committed:
[TAN-4519] Set `supports_public_visibility?` to true in CommonGround

adessy committed:
[TAN-4519] Refactor `transitive` condition in IdeasFinder

The finder can now be used to retrieve both transitive and non-transitive ideas, whereas previously it only supported transitive=true case. Additionally, the parameter is now explicitly converted into a boolean using `Utils.to_bool`, so string values like ``"false"` are correctly treated as `false`.

adessy committed:
[TAN-4519] Speed up `IdeasFinder` specs

adessy committed:
[TAN-4519] Disable auto-reactions on idea submission for some participation methods

With this change, auto-reactions are only enabled for proposals and ideation participation methods.

adessy committed:
[TAN-4519] Fix broken tests

adessy committed:
[TAN-4519] Test idea updates in common ground phases

adessy commented:
@jamesspeake, ping me if you want some more context about this one.

adessy commented:
Hey @jamesspeake, I’ve removed the review request for now. I realized there were a few things missing.

adessy committed:
Fix typo

adessy committed:
Merge branch 'master' into tan-4519/minimal-common-ground-participation-method-be

adessy committed:
[TAN-4519] Add neutral reactions

Ideas can now receive up, down, and neutral reactions. For now, neutral reactions will be used to support the "pass/ignore" option in the Common Ground participation method.

adessy committed:
[TAN-4519] Add test for `neutral_reactions_count` on ideas

adessy committed:
[TAN-4519] Fix rubocop offense

adessy committed:
[TAN-4519] Remove references to `answer_visible_to` and `CustomField::VISIBLE_TO_PUBLIC`

Those things were removed in a concurrent refactoring of `CustomField`s.

adessy committed:
[TAN-4519] Remove TODO comment

adessy committed:
[TAN-4519] Speed up `IdeasFinder` tests

adessy committed:
Fix flaky test

adessy committed:
Merge branch 'master' into tan-4519/minimal-common-ground-participation-method-be

# Conflicts:
#	back/app/models/idea.rb
#	back/app/services/idea_custom_fields_service.rb
#	back/spec/lib/participation_method/ideation_spec.rb
#	back/spec/lib/participation_method/information_spec.rb
#	back/spec/lib/participation_method/native_survey_spec.rb
#	back/spec/lib/participation_method/none_spec.rb
#	back/spec/lib/participation_method/poll_spec.rb
#	back/spec/lib/participation_method/proposals_spec.rb
#	back/spec/lib/participation_method/survey_spec.rb
#	back/spec/lib/participation_method/volunteering_spec.rb
#	back/spec/lib/participation_method/voting_spec.rb
#	back/spec/models/phase_spec.rb

sebastienhoorens commented on back/lib/participation_method/base.rb:
```diff
@@ -194,6 +214,22 @@ def supports_multiple_phase_reports?
+    # See +Idea::SUBMISSION_STATUSES+ for what it means for an input to be considered
+    # submitted.
+    #
+    # One issue with this is that reactions aren’t scoped to a specific phase, and an
```
About the comments on phases: I think that ideally it would be the creation phase that would determine this behaviour, especially as submission (always?) happens in the creation phase. Today, ideation ideas don't have a creation phase, but I think Koen's feedback was that in the long run also ideation ideas should get a creation phase.

adessy replied:
Agreed, but as you mentioned, we currently don't persist the creation phase for all ideas. In practice, the current implementationkinda has the same behavior, since an input should, in principle, only be associated with its creation phase before its publication or submission.

sebastienhoorens commented on back/lib/participation_method/common_ground.rb:
```diff
@@ -0,0 +1,57 @@
+      input.idea_status ||= IdeaStatus.find_by!(code: 'proposed', participation_method: 'ideation')
+    end
+
+    # This is necessary in order to be able to retrieve the inputs using the +IdeasFinder+
```
Seems acceptable to me.

sebastienhoorens commented on back/lib/participation_method/common_ground.rb:
```diff
@@ -0,0 +1,57 @@
+      multiloc_service = MultilocService.new
+
+      [
+        CustomField.new(
```
Is there a reason to not have a `description_multiloc`? If not, then perhaps it could be useful to have a method `default_title_multiloc_field` or so that it remains consistent between ideation and common ground. If yes, then I think it could be useful to add a comment.

adessy replied:
Inputs in a common ground phase are essentially short statements, so we have no use for a body at the moment. We'll probably reconsider this depending on what comes out of user feedback and usage. Will add a comment.

sebastienhoorens commented on back/spec/finders/ideas_finder_spec.rb:
```diff
@@ -43,13 +44,21 @@
   end
 
   describe '#transitive_condition' do
```
I don't see a common ground input being created in this spec. Do you think it could be useful to add one and check when it should and shouldn't be returned?

adessy replied:
I modified this file because I fixed a small bug in the `transitive` filter of the finder. There is no filter specific to common ground inputs, and common ground inputs are no different from ideation inputs, so I think we're covered.

sebastienhoorens commented on back/spec/models/reaction_spec.rb:
```diff
@@ -14,7 +14,30 @@
   end
 
-  context 'uniquness' do
+  context 'Neutral Reaction' do
```
I think it could be useful to also add an acceptance test for voting (reacting) on a common ground.

sebastienhoorens reviewed (approved):
Great feature! 👍 Do we need a feature flag for it?

I added a few comments, but mostly nitpicking and nothing blocking. Also, just checking: will emails (like idea published or user digest) work as desired?

adessy committed:
[TAN-4519] Add clarification comment for input fields in CommonGround

adessy committed:
[TAN-4519] Add acceptance test for adding a neutral reaction to an idea

adessy committed:
[TAN-4519] Fix rubocop offense

adessy committed:
[TAN-4519] Fix broken idea tests

adessy committed:
[TAN-4519] Enable dislike reactions by default for common ground phases

Like (up) and neutral reactions are always allowed. For consistency, we could add another flag to enable or disable neutral reactions, but it's not strictly necessary, as this can be implicitly inferred by the front end based on the participation method (which is not the case for dislike (down) reactions).

adessy committed:
[TAN-4519] Fix Common Ground test

The `supports_multiple_posts` method has been replaced by `allow_posting_again_after` in a concurrent PR.
